source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,598 @@
+          {SPV_OPERAND_TYPE_LITERAL_INTEGER, {0}}}));
+  ir::Instruction* integer = &*--context()->types_values_end();
+  get_def_use_mgr()->AnalyzeInstDef(integer);
+
+  return id;
+}
+
+void ScalarReplacementPass::GetOrCreateInitialValue(ir::Instruction* source,
+                                                    uint32_t index,
+                                                    ir::Instruction* newVar) {
+  assert(source->opcode() == SpvOpVariable);
+  if (source->NumInOperands() < 2) return;
+
+  uint32_t initId = source->GetSingleWordInOperand(1u);
+  uint32_t storageId = GetStorageType(newVar)->result_id();
+  ir::Instruction* init = get_def_use_mgr()->GetDef(initId);
+  uint32_t newInitId = 0;
+  if (init->opcode() == SpvOpConstantNull) {
+    // Initialize to appropriate NULL.
+    auto iter = type_to_null_.find(storageId);
+    if (iter == type_to_null_.end()) {
+      newInitId = TakeNextId();
+      type_to_null_[storageId] = newInitId;
+      context()->AddGlobalValue(MakeUnique<ir::Instruction>(
+          context(), SpvOpConstantNull, storageId, newInitId,
+          std::initializer_list<ir::Operand>{}));
+      ir::Instruction* newNull = &*--context()->types_values_end();
+      get_def_use_mgr()->AnalyzeInstDefUse(newNull);
+    } else {
+      newInitId = iter->second;
+    }
+  } else if (ir::IsSpecConstantInst(init->opcode())) {
+    // Generate a constant for the extraction.
+    uint32_t constantId = 0;
+    auto iter = index_to_constant_.find(index);
+    if (iter == index_to_constant_.end()) {
+      if (int_id_ == 0) {
+        int_id_ = GetIntId();
+      }
+      constantId = TakeNextId();
+      context()->AddGlobalValue(MakeUnique<ir::Instruction>(
+          context(), SpvOpConstant, int_id_, constantId,
+          std::initializer_list<ir::Operand>{
+              {SPV_OPERAND_TYPE_LITERAL_INTEGER, {index}}}));
+      ir::Instruction* constant = &*--context()->types_values_end();
+      get_def_use_mgr()->AnalyzeInstDefUse(constant);
+    } else {
+      constantId = iter->second;
+    }
+
+    // Create a new constant extract.
+    newInitId = TakeNextId();
+    context()->AddGlobalValue(MakeUnique<ir::Instruction>(
+        context(), SpvOpSpecConstantOp, storageId, newInitId,
+        std::initializer_list<ir::Operand>{
+            {SPV_OPERAND_TYPE_SPEC_CONSTANT_OP_NUMBER, {SpvOpCompositeExtract}},
+            {SPV_OPERAND_TYPE_ID, {constantId}}}));
+    ir::Instruction* newSpecConst = &*--context()->types_values_end();
+    get_def_use_mgr()->AnalyzeInstDefUse(newSpecConst);
+  } else if (init->opcode() == SpvOpConstantComposite) {
+    // Get the appropriate index constant.
+    newInitId = init->GetSingleWordInOperand(index);
+    ir::Instruction* element = get_def_use_mgr()->GetDef(newInitId);
+    if (element->opcode() == SpvOpUndef) {
+      // Undef is not a valid initializer for a variable.
+      newInitId = 0;
+    }
+  } else {
+    assert(false);
```

Comment:
The spec specifically states that the initializer could be the id of a global variable.  Add a comment that you are specifically disallowing it.  Also I don't see the code in |CanReplaceVariable| that disallows this case.  Should be added there.

Reply:
Since OpVariable returns a pointer, I won't see any valid cases for scalarization (only composite types are handled).